### ISSUE TITLE
potential support for browser history

### ISSUE THREAD

**Issue by AllanHoejgaardJensen** [michaelbromley/angularUtils#119]
Hi,
Great work first of all, and this makes pagination simpler, which is good.

I do have an issue or more correct a question. 

I was wondering if support for browser history is possible in any other way than using the "on-page-change="handleHistory(newPageNumber)" and use $location.hash for paging numbers and $location.search for the 'filtering value' . This allows a user to eg. return to the state of the pagination on a page previously visited, using the back button. Do you have a recommendation for handling this kind of issue?

Do you think it would be a good thing to have support for browser history as a build in feature of the Pagination eg. as a property in the 'dir-pagination-controls' eg. "history" having values: default = 'no support', or 'filter', 'page' and 'full'. And then a page reload strategy default of reloadOnSearch=false for $routeProvider.

Thank you very much for your time.

**Comment by michaelbromley** [comment]
Hi,

There are probably many ways to implement persistence of the pagination state - using the window.location as you suggest is probably the most straightforward and best, since it produces deep-linkable results. Other approaches could be to use cookies or localStorage to store the various parameters (currentPage, filters, ordering etc).

For example, here is how I did it in one of my own apps where I use this pagination module: https://github.com/michaelbromley/drawACatApp/blob/master/src/app/home/home.js#L61

As you can see, I use the `on-page-change` callback to update the location, but rather than manually updating the `$location` object, I let ui-router's `$state` service do it for me. But essentially it is the same idea as what you suggest in your question.

Regarding building this kind of functionality into the module itself, I would not be inclined to do so. The reason is that the best scheme of storing the state in the URL is highly dependent on each individual project's environment. For example, you suggest storing page number in the "hash" part and filter values in the "search" part. This scheme might well break depending on whether or not a project uses html5Mode for the URLs. Or a project may already use hashes to denote something else already. Or - as in the case of my project above, I use `$stateProvider` rather than the built-in `$routeProvider`. 

In short, I'd have to make too many assumptions about things which should not be the concern of a mere pagination directive. 

Unless, of course, you have an idea for a way to get around all these issues and have something that would work independent of the wider project's environment settings. In that case I'd be happy to hear it and then see if I could figure out an implementation.

**Comment by seanrond** [comment]
@michaelbromley I was planning to implement the persistence of the pagination state as you described here. The part I am missing is how you hook into the browser back/forward events and how to update the pagination state programatically.

Could you point to where this is done in your example please?

**Comment by michaelbromley** [comment]
@seanrond you could perhaps use event listeners on the route/state change success event (specific to the router you are using).

Another way could be to create a $watch on the `$location` service, e.g. `$location.url()` together with `$location.search()`.
